# Table of Contents

Questions:

01. Import Libraries and Data.
02. Addressing PII Data.
03. Creating Regional Segmentation of the Data and determine differences in spending habits between the different U.S. regions.
04. Creating an exclusion flag for low-activity customers (customers with less than 5 orders) and excluding them from the data.


# 01. Import Libraries and Data

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns 
import scipy 

In [2]:
# Set path
path = r'C:\Users\Goce\Desktop\Achievement 4 Python Fundamentals'

In [3]:
# import data 
ords_prods_cust_merged = pd.read_pickle(os.path.join( path, 'Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))

In [4]:
# Set the max_columns option to None
pd.set_option('display.max_columns', None)

In [5]:
#checking the top of the data 
ords_prods_cust_merged.head()

user_id  gender     state  age date_joined  number_of_dependants  \
0    26711  Female  Missouri   48    1/1/2017                     3   
1    26711  Female  Missouri   48    1/1/2017                     3   
2    26711  Female  Missouri   48    1/1/2017                     3   
3    26711  Female  Missouri   48    1/1/2017                     3   
4    26711  Female  Missouri   48    1/1/2017                     3   

  marital_status  income  order_id  order_number  orders_day_of_week  \
0        married  165665   2543867             5                   1   
1        married  165665   1285508             7                   5   
2        married  165665   2578584             8                   1   
3        married  165665    423547             2                   2   
4        married  165665   2524893             3                   3   

   order_hour_of_day  days_since_last-order  new_customer  product_id  \
0                  9                   30.0         False         196   
1                 15                   11.0         False         196   
2                 15                   10.0         False         196   
3                  9                   14.0         False       38928   
4                 11                   30.0         False       38928   

   add_to_cart_order  reordered              product_name  aisle_id  \
0                  2          0                      Soda        77   
1                  1          1                      Soda        77   
2                  2          1                      Soda        77   
3                  1          0  0% Greek Strained Yogurt       120   
4                  1          1  0% Greek Strained Yogurt       120   

   department_id     prices    price_range_loc     busiest_day  \
0              7   9.000000  Mid-range product  Regularly busy   
1              7   9.000000  Mid-range product  Regularly busy   
2              7   9.000000  Mid-range product  Regularly busy   
3             16  12.601562  Mid-range product  Regularly busy   
4             16  12.601562  Mid-range product  Regularly busy   

     busiest_days busiest_period_of_day  max_order  loyalty_flag  \
0    Busiest days           Most orders          8  New customer   
1  Regularly busy           Most orders          8  New customer   
2    Busiest days           Most orders          8  New customer   
3  Regularly busy           Most orders          8  New customer   
4    Slowest days           Most orders          8  New customer   

   user_avg_price spending_flag  median_prior_orders order_frequency_flag  \
0        7.988281   Low spender                 19.0     Regular customer   
1        7.988281   Low spender                 19.0     Regular customer   
2        7.988281   Low spender                 19.0     Regular customer   
3        7.988281   Low spender                 19.0     Regular customer   
4        7.988281   Low spender                 19.0     Regular customer   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [6]:
#checking the bottom of the data 
ords_prods_cust_merged.tail()

user_id  gender     state  age date_joined  number_of_dependants  \
32404854    80148  Female  New York   55    4/1/2020                     1   
32404855    80148  Female  New York   55    4/1/2020                     1   
32404856    80148  Female  New York   55    4/1/2020                     1   
32404857    80148  Female  New York   55    4/1/2020                     1   
32404858    80148  Female  New York   55    4/1/2020                     1   

         marital_status  income  order_id  order_number  orders_day_of_week  \
32404854        married   57095   2859858             3                   6   
32404855        married   57095   2859858             3                   6   
32404856        married   57095   3209855             4                   6   
32404857        married   57095   2859858             3                   6   
32404858        married   57095   2859858             3                   6   

          order_hour_of_day  days_since_last-order  new_customer  product_id  \
32404854                 11                   12.0         False       38750   
32404855                 11                   12.0         False       39972   
32404856                 19                    7.0         False        3198   
32404857                 11                   12.0         False       39069   
32404858                 11                   12.0         False       21167   

          add_to_cart_order  reordered  \
32404854                 11          0   
32404855                  5          0   
32404856                  2          0   
32404857                  2          0   
32404858                  1          0   

                                               product_name  aisle_id  \
32404854                     Protein Bar, Chocolate Pretzel         3   
32404855                   Chocolate Fudge High Protein Bar         3   
32404856  Pumpkin Spice Latte Iced Coffee Pure Cold Brew...        26   
32404857               Chocolate Coconut Almond Protein Bar         3   
32404858                         Lemon Vanilla Luna Protein         3   

          department_id    prices    price_range_loc     busiest_day  \
32404854             19  4.699219  Low-range product  Regularly busy   
32404855             19  4.000000  Low-range product  Regularly busy   
32404856              7  4.699219  Low-range product  Regularly busy   
32404857             19  1.799805  Low-range product  Regularly busy   
32404858             19  6.000000  Mid-range product  Regularly busy   

            busiest_days busiest_period_of_day  max_order  loyalty_flag  \
32404854  Regularly busy           Most orders          4  New customer   
32404855  Regularly busy           Most orders          4  New customer   
32404856  Regularly busy        Average orders          4  New customer   
32404857  Regularly busy           Most orders          4  New customer   
32404858  Regularly busy           Most orders          4  New customer   

          user_avg_price spending_flag  median_prior_orders  \
32404854        3.886719   Low spender                 12.0   
32404855        3.886719   Low spender                 12.0   
32404856        3.886719   Low spender                 12.0   
32404857        3.886719   Low spender                 12.0   
32404858        3.886719   Low spender                 12.0   

         order_frequency_flag _merge  
32404854     Regular customer   both  
32404855     Regular customer   both  
32404856     Regular customer   both  
32404857     Regular customer   both  
32404858     Regular customer   both

# 02. Addressing PII Data 

Sensitive customer information (Names and Surnames) are excluded in the previuos step of the analysis (4.9 Intro to data visualization with Python Part 1).

In [7]:
# checking dataframe information
ords_prods_cust_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   user_id                int32   
 1   gender                 object  
 2   state                  object  
 3   age                    int8    
 4   date_joined            object  
 5   number_of_dependants   int8    
 6   marital_status         object  
 7   income                 int64   
 8   order_id               int32   
 9   order_number           int8    
 10  orders_day_of_week     int8    
 11  order_hour_of_day      int8    
 12  days_since_last-order  float16 
 13  new_customer           bool    
 14  product_id             int32   
 15  add_to_cart_order      int32   
 16  reordered              int8    
 17  product_name           object  
 18  aisle_id               int8    
 19  department_id          int8    
 20  prices                 float16 
 21  price_range_loc        object

# 03. Creating Regional Segmentation of the Data and determine differences in spending habits between the different U.S. regions.

In [8]:
# Defining Regions
Northeast = ['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Rhode Island', 'Connecticut', 'New York', 'Pennsylvania', 'New Jersey']
Midwest = ['Wisconsin', 'Michigan','Illinois', 'Indiana', 'Ohio', 'North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 'Missouri']
South = ['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia', 'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee', 'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']
West = ['Idaho', 'Montana', 'Wyoming', 'Nevada', 'Utah', 'Colorado', 'Arizona', 'New Mexico', 'Alaska', 'Washington', 'Oregon', 'California', 'Hawaii']

In [9]:
# Creating region column 
ords_prods_cust_merged.loc[ords_prods_cust_merged['state'].isin (Northeast), 'region'] = 'Northeast'
ords_prods_cust_merged.loc[ords_prods_cust_merged['state'].isin (Midwest), 'region'] = 'Midwest'
ords_prods_cust_merged.loc[ords_prods_cust_merged['state'].isin (South), 'region'] = 'South'
ords_prods_cust_merged.loc[ords_prods_cust_merged['state'].isin (West), 'region'] = 'West'

In [10]:
# checking the shape of dataframe
ords_prods_cust_merged.shape

(32404859, 33)

In [11]:
# checking dataframe information
ords_prods_cust_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   user_id                int32   
 1   gender                 object  
 2   state                  object  
 3   age                    int8    
 4   date_joined            object  
 5   number_of_dependants   int8    
 6   marital_status         object  
 7   income                 int64   
 8   order_id               int32   
 9   order_number           int8    
 10  orders_day_of_week     int8    
 11  order_hour_of_day      int8    
 12  days_since_last-order  float16 
 13  new_customer           bool    
 14  product_id             int32   
 15  add_to_cart_order      int32   
 16  reordered              int8    
 17  product_name           object  
 18  aisle_id               int8    
 19  department_id          int8    
 20  prices                 float16 
 21  price_range_loc        object

In [12]:
ords_prods_cust_merged['region'].value_counts()

South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: region, dtype: int64

In [20]:
# in order to count each user only once
ords_prods_cust_merged.groupby(['region']).user_id.nunique()

region
Midwest      48519
Northeast    36388
South        68737
West         52565
Name: user_id, dtype: int64

In [13]:
# Checking new column
ords_prods_cust_merged.head(20)

user_id  gender       state  age date_joined  number_of_dependants  \
0     26711  Female    Missouri   48    1/1/2017                     3   
1     26711  Female    Missouri   48    1/1/2017                     3   
2     26711  Female    Missouri   48    1/1/2017                     3   
3     26711  Female    Missouri   48    1/1/2017                     3   
4     26711  Female    Missouri   48    1/1/2017                     3   
5     26711  Female    Missouri   48    1/1/2017                     3   
6     26711  Female    Missouri   48    1/1/2017                     3   
7     26711  Female    Missouri   48    1/1/2017                     3   
8     26711  Female    Missouri   48    1/1/2017                     3   
9     26711  Female    Missouri   48    1/1/2017                     3   
10    26711  Female    Missouri   48    1/1/2017                     3   
11    26711  Female    Missouri   48    1/1/2017                     3   
12    26711  Female    Missouri   48    1/1/2017                     3   
13    26711  Female    Missouri   48    1/1/2017                     3   
14    26711  Female    Missouri   48    1/1/2017                     3   
15    26711  Female    Missouri   48    1/1/2017                     3   
16    26711  Female    Missouri   48    1/1/2017                     3   
17    26711  Female    Missouri   48    1/1/2017                     3   
18    33890  Female  New Mexico   36    1/1/2017                     0   
19    33890  Female  New Mexico   36    1/1/2017                     0   

   marital_status  income  order_id  order_number  orders_day_of_week  \
0         married  165665   2543867             5                   1   
1         married  165665   1285508             7                   5   
2         married  165665   2578584             8                   1   
3         married  165665    423547             2                   2   
4         married  165665   2524893             3                   3   
5         married  165665   2984525             4                   1   
6         married  165665   2505178             6                   1   
7         married  165665   1285508             7                   5   
8         married  165665   2578584             8                   1   
9         married  165665    518967             1                   2   
10        married  165665   2524893             3                   3   
11        married  165665   2984525             4                   1   
12        married  165665   2543867             5                   1   
13        married  165665   2505178             6                   1   
14        married  165665   1285508             7                   5   
15        married  165665   2578584             8                   1   
16        married  165665   2524893             3                   3   
17        married  165665   2543867             5                   1   
18         single   59285   1044841             3                   6   
19         single   59285    474720             4                   1   

    order_hour_of_day  days_since_last-order  new_customer  product_id  \
0                   9                   30.0         False         196   
1                  15                   11.0         False         196   
2                  15                   10.0         False         196   
3                   9                   14.0         False       38928   
4                  11                   30.0         False       38928   
5                  17                   19.0         False       38928   
6                   8                   21.0         False       38928   
7                  15                   11.0         False       38928   
8                  15                   10.0         False       38928   
9                   9                    NaN          True        6184   
10                 11                   30.0         False        6184   
11                 17      

In [14]:
# Checking new column
ords_prods_cust_merged.tail(20)

user_id  gender       state  age date_joined  number_of_dependants  \
32404839    81095  Female  California   27    4/1/2020                     1   
32404840    81095  Female  California   27    4/1/2020                     1   
32404841    81095  Female  California   27    4/1/2020                     1   
32404842    81095  Female  California   27    4/1/2020                     1   
32404843    81095  Female  California   27    4/1/2020                     1   
32404844    80148  Female    New York   55    4/1/2020                     1   
32404845    80148  Female    New York   55    4/1/2020                     1   
32404846    80148  Female    New York   55    4/1/2020                     1   
32404847    80148  Female    New York   55    4/1/2020                     1   
32404848    80148  Female    New York   55    4/1/2020                     1   
32404849    80148  Female    New York   55    4/1/2020                     1   
32404850    80148  Female    New York   55    4/1/2020                     1   
32404851    80148  Female    New York   55    4/1/2020                     1   
32404852    80148  Female    New York   55    4/1/2020                     1   
32404853    80148  Female    New York   55    4/1/2020                     1   
32404854    80148  Female    New York   55    4/1/2020                     1   
32404855    80148  Female    New York   55    4/1/2020                     1   
32404856    80148  Female    New York   55    4/1/2020                     1   
32404857    80148  Female    New York   55    4/1/2020                     1   
32404858    80148  Female    New York   55    4/1/2020                     1   

         marital_status  income  order_id  order_number  orders_day_of_week  \
32404839        married   99799    426229             3                   0   
32404840        married   99799    426229             3                   0   
32404841        married   99799   2855401             2                   4   
32404842        married   99799   2855401             2                   4   
32404843        married   99799   2855401             2                   4   
32404844        married   57095   2859858             3                   6   
32404845        married   57095   2859858             3                   6   
32404846        married   57095   3209855             4                   6   
32404847        married   57095   2859858             3                   6   
32404848        married   57095   2859858             3                   6   
32404849        married   57095   2028188             1                   6   
32404850        married   57095   1393131             2                   1   
32404851        married   57095   2859858             3                   6   
32404852        married   57095   2859858             3                   6   
32404853        married   57095   2859858             3                   6   
32404854        married   57095   2859858             3                   6   
32404855        married   57095   2859858             3                   6   
32404856        married   57095   3209855             4                   6   
32404857        married   57095   2859858             3                   6   
32404858        married   57095   2859858             3                   6   

          order_hour_of_day  days_since_last-order  new_customer  product_id  \
32404839                 18                   30.0         False       32403   
32404840                 18                   30.0         False       45588   
32404841                 19                   22.0         False        3438   
32404842                 19                   22.0         False       38406   
32404843                 19                   22.0         False       15024   
32404844                 11                   12.0         False       36554   
32404845                 11                   12.0         False       48335   
32404846                 19  

In [24]:
# Checking spending habits by regions
spending_habits2 = pd.crosstab(ords_prods_cust_merged['region'], ords_prods_cust_merged['spending_flag'], dropna = False)

In [25]:
# Spending habits in different regions
spending_habits2

spending_flag  High spender  Low spender
region                                  
Midwest               29376      7567949
Northeast             18642      5704094
South                 40739     10751146
West                  31366      8261547

In [23]:
# in order to count each user only once
spending_habits = pd.crosstab(ords_prods_cust_merged['region'], ords_prods_cust_merged['spending_flag'], dropna = False, values=ords_prods_cust_merged['user_id'], 
                              aggfunc='nunique')

In [26]:
#spending habits by regions and individual users
spending_habits

spending_flag  High spender  Low spender
region                                  
Midwest                 803        47716
Northeast               545        35843
South                  1144        67593
West                    904        51661

# 04. Creating an exclusion flag for low-activity customers (customers with less than 5 orders) and excluding them from the data

In [27]:
# To create activity flag, we are using the following criteria:
# if the customer has less than 5 orders, then the customer should be labeled as “low active customer.”
# If the customer has 5 or more orders, then the customer should be labeled as “high active customer.”

In [28]:
# Creating flag for customers with 5 or more orders.
ords_prods_cust_merged.loc[ords_prods_cust_merged['max_order'] >= 5, 'customer_activity'] = 'High active'

In [29]:
# Creating flag for customers with less than 5 orders
ords_prods_cust_merged.loc[ords_prods_cust_merged['max_order'] <5, 'customer_activity'] = 'Low active'

In [30]:
#checking the shape
ords_prods_cust_merged.shape

(32404859, 34)

In [31]:
#checking the values in customer activity 
ords_prods_cust_merged['customer_activity'].value_counts(dropna = False)

High active    30964564
Low active      1440295
Name: customer_activity, dtype: int64

In [32]:
# in order to count each user only once
ords_prods_cust_merged.groupby(['customer_activity']).user_id.nunique()

customer_activity
High active    162631
Low active      43578
Name: user_id, dtype: int64

This means that we will remove 43578 users from further analysis

In [35]:
#defining the dataframe with low active customers/creating subset
low_active_customers = ords_prods_cust_merged[ords_prods_cust_merged['customer_activity'] == 'Low active']

In [36]:
low_active_customers.head()

user_id gender           state  age date_joined  number_of_dependants  \
729   168851   Male  South Carolina   30    1/1/2017                     0   
730   168851   Male  South Carolina   30    1/1/2017                     0   
731   168851   Male  South Carolina   30    1/1/2017                     0   
732   168851   Male  South Carolina   30    1/1/2017                     0   
733   168851   Male  South Carolina   30    1/1/2017                     0   

    marital_status  income  order_id  order_number  orders_day_of_week  \
729         single   63712   1903574             1                   2   
730         single   63712   1193140             2                   3   
731         single   63712   1870356             3                   1   
732         single   63712   1903574             1                   2   
733         single   63712   1193140             2                   3   

     order_hour_of_day  days_since_last-order  new_customer  product_id  \
729                  8                    NaN          True         196   
730                 14                   15.0         False         196   
731                  8                   26.0         False        6184   
732                  8                    NaN          True         130   
733                 14                   15.0         False         130   

     add_to_cart_order  reordered  \
729                  2          0   
730                  1          1   
731                  2          0   
732                  3          0   
733                  2          1   

                                          product_name  aisle_id  \
729                                               Soda        77   
730                                               Soda        77   
731                                        Clementines        32   
732  Vanilla Milk Chocolate Almond Ice Cream Bars M...        37   
733  Vanilla Milk Chocolate Almond Ice Cream Bars M...        37   

     department_id    prices    price_range_loc     busiest_day  \
729              7  9.000000  Mid-range product  Regularly busy   
730              7  9.000000  Mid-range product  Regularly busy   
731              4  4.300781  Low-range product  Regularly busy   
732              1  8.203125  Mid-range product  Regularly busy   
733              1  8.203125  Mid-range product  Regularly busy   

       busiest_days busiest_period_of_day  max_order  loyalty_flag  \
729  Regularly busy        Average orders          3  New customer   
730    Slowest days           Most orders          3  New customer   
731    Busiest days        Average orders          3  New customer   
732  Regularly busy        Average orders          3  New customer   
733    Slowest days           Most orders          3  New customer   

     user_avg_price spending_flag  median_prior_orders   order_frequency_flag  \
729        7.488281   Low spender                 20.5  Non-frequent customer   
730        7.488281   Low spender                 20.5  Non-frequent customer   
731        7.488281   Low spender                 20.5  Non-frequent customer   
732        7.488281   Low spender                 20.5  Non-frequent customer   
733        7.488281   Low spender                 20.5  Non-frequent customer   

    _merge region customer_activity  
729   both  South        Low active  
730   both  South        Low active  
731   both  South        Low active  
732   both  South        Low active  
733   both  South        Low active

In [37]:
#checking the bottom
low_active_customers.tail()

user_id  gender     state  age date_joined  number_of_dependants  \
32404854    80148  Female  New York   55    4/1/2020                     1   
32404855    80148  Female  New York   55    4/1/2020                     1   
32404856    80148  Female  New York   55    4/1/2020                     1   
32404857    80148  Female  New York   55    4/1/2020                     1   
32404858    80148  Female  New York   55    4/1/2020                     1   

         marital_status  income  order_id  order_number  orders_day_of_week  \
32404854        married   57095   2859858             3                   6   
32404855        married   57095   2859858             3                   6   
32404856        married   57095   3209855             4                   6   
32404857        married   57095   2859858             3                   6   
32404858        married   57095   2859858             3                   6   

          order_hour_of_day  days_since_last-order  new_customer  product_id  \
32404854                 11                   12.0         False       38750   
32404855                 11                   12.0         False       39972   
32404856                 19                    7.0         False        3198   
32404857                 11                   12.0         False       39069   
32404858                 11                   12.0         False       21167   

          add_to_cart_order  reordered  \
32404854                 11          0   
32404855                  5          0   
32404856                  2          0   
32404857                  2          0   
32404858                  1          0   

                                               product_name  aisle_id  \
32404854                     Protein Bar, Chocolate Pretzel         3   
32404855                   Chocolate Fudge High Protein Bar         3   
32404856  Pumpkin Spice Latte Iced Coffee Pure Cold Brew...        26   
32404857               Chocolate Coconut Almond Protein Bar         3   
32404858                         Lemon Vanilla Luna Protein         3   

          department_id    prices    price_range_loc     busiest_day  \
32404854             19  4.699219  Low-range product  Regularly busy   
32404855             19  4.000000  Low-range product  Regularly busy   
32404856              7  4.699219  Low-range product  Regularly busy   
32404857             19  1.799805  Low-range product  Regularly busy   
32404858             19  6.000000  Mid-range product  Regularly busy   

            busiest_days busiest_period_of_day  max_order  loyalty_flag  \
32404854  Regularly busy           Most orders          4  New customer   
32404855  Regularly busy           Most orders          4  New customer   
32404856  Regularly busy        Average orders          4  New customer   
32404857  Regularly busy           Most orders          4  New customer   
32404858  Regularly busy           Most orders          4  New customer   

          user_avg_price spending_flag  median_prior_orders  \
32404854        3.886719   Low spender                 12.0   
32404855        3.886719   Low spender                 12.0   
32404856        3.886719   Low spender                 12.0   
32404857        3.886719   Low spender                 12.0   
32404858        3.886719   Low spender                 12.0   

         order_frequency_flag _merge     region customer_activity  
32404854     Regular customer   both  Northeast        Low active  
32404855     Regular customer   both  Northeast        Low active  
32404856     Regular customer   both  Northeast        Low active  
32404857     Regular customer   both  Northeast        Low active  
32404858     Regular customer   both  Northeast        Low active

In [38]:
low_active_customers.shape

(1440295, 34)

In [39]:
# Exporting subset 'low_active_customers'
low_active_customers.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'Low_active_customers_less_than_5_orders.pkl'))

In [45]:
#defining the dataframe with high active customers
high_active_customers = ords_prods_cust_merged[ords_prods_cust_merged['customer_activity'] == 'High active']

In [46]:
#checking the top
high_active_customers.head()

user_id  gender     state  age date_joined  number_of_dependants  \
0    26711  Female  Missouri   48    1/1/2017                     3   
1    26711  Female  Missouri   48    1/1/2017                     3   
2    26711  Female  Missouri   48    1/1/2017                     3   
3    26711  Female  Missouri   48    1/1/2017                     3   
4    26711  Female  Missouri   48    1/1/2017                     3   

  marital_status  income  order_id  order_number  orders_day_of_week  \
0        married  165665   2543867             5                   1   
1        married  165665   1285508             7                   5   
2        married  165665   2578584             8                   1   
3        married  165665    423547             2                   2   
4        married  165665   2524893             3                   3   

   order_hour_of_day  days_since_last-order  new_customer  product_id  \
0                  9                   30.0         False         196   
1                 15                   11.0         False         196   
2                 15                   10.0         False         196   
3                  9                   14.0         False       38928   
4                 11                   30.0         False       38928   

   add_to_cart_order  reordered              product_name  aisle_id  \
0                  2          0                      Soda        77   
1                  1          1                      Soda        77   
2                  2          1                      Soda        77   
3                  1          0  0% Greek Strained Yogurt       120   
4                  1          1  0% Greek Strained Yogurt       120   

   department_id     prices    price_range_loc     busiest_day  \
0              7   9.000000  Mid-range product  Regularly busy   
1              7   9.000000  Mid-range product  Regularly busy   
2              7   9.000000  Mid-range product  Regularly busy   
3             16  12.601562  Mid-range product  Regularly busy   
4             16  12.601562  Mid-range product  Regularly busy   

     busiest_days busiest_period_of_day  max_order  loyalty_flag  \
0    Busiest days           Most orders          8  New customer   
1  Regularly busy           Most orders          8  New customer   
2    Busiest days           Most orders          8  New customer   
3  Regularly busy           Most orders          8  New customer   
4    Slowest days           Most orders          8  New customer   

   user_avg_price spending_flag  median_prior_orders order_frequency_flag  \
0        7.988281   Low spender                 19.0     Regular customer   
1        7.988281   Low spender                 19.0     Regular customer   
2        7.988281   Low spender                 19.0     Regular customer   
3        7.988281   Low spender                 19.0     Regular customer   
4        7.988281   Low spender                 19.0     Regular customer   

  _merge   region customer_activity  
0   both  Midwest       High active  
1   both  Midwest       High active  
2   both  Midwest       High active  
3   both  Midwest       High active  
4   both  Midwest       High active

In [47]:
#checking the bottom
high_active_customers.tail()

user_id  gender     state  age date_joined  number_of_dependants  \
32404813   135902  Female  Missouri   66    4/1/2020                     2   
32404814   135902  Female  Missouri   66    4/1/2020                     2   
32404815   135902  Female  Missouri   66    4/1/2020                     2   
32404816   135902  Female  Missouri   66    4/1/2020                     2   
32404817   135902  Female  Missouri   66    4/1/2020                     2   

         marital_status  income  order_id  order_number  orders_day_of_week  \
32404813        married   53374    974951             1                   6   
32404814        married   53374    628984             4                   5   
32404815        married   53374   1157983             7                   5   
32404816        married   53374   2498530             8                   6   
32404817        married   53374   1104840             3                   6   

          order_hour_of_day  days_since_last-order  new_customer  product_id  \
32404813                 17                    NaN          True       31136   
32404814                 14                   13.0         False       31136   
32404815                 15                    7.0         False       31136   
32404816                 21                    8.0         False       32312   
32404817                 11                   23.0         False        4909   

          add_to_cart_order  reordered  \
32404813                  1          0   
32404814                  9          1   
32404815                  2          1   
32404816                  7          0   
32404817                  5          0   

                                      product_name  aisle_id  department_id  \
32404813       Original Flavor Sugar Free Gum 3 Pk        46             19   
32404814       Original Flavor Sugar Free Gum 3 Pk        46             19   
32404815       Original Flavor Sugar Free Gum 3 Pk        46             19   
32404816  Glide Pro-Health Comfort Plus Mint Floss        20             11   
32404817              Sweet Hot Mustard with Honey        72             13   

            prices    price_range_loc     busiest_day    busiest_days  \
32404813  4.898438  Low-range product  Regularly busy  Regularly busy   
32404814  4.898438  Low-range product  Regularly busy  Regularly busy   
32404815  4.898438  Low-range product  Regularly busy  Regularly busy   
32404816  8.398438  Mid-range product  Regularly busy  Regularly busy   
32404817  7.699219  Mid-range product  Regularly busy  Regularly busy   

         busiest_period_of_day  max_order  loyalty_flag  user_avg_price  \
32404813        Average orders          8  New customer        6.960938   
32404814           Most orders          8  New customer        6.960938   
32404815           Most orders          8  New customer        6.960938   
32404816        Average orders          8  New customer        6.960938   
32404817           Most orders          8  New customer        6.960938   

         spending_flag  median_prior_orders order_frequency_flag _merge  \
32404813   Low spender                  8.0    Frequent customer   both   
32404814   Low spender                  8.0    Frequent customer   both   
32404815   Low spender                  8.0    Frequent customer   both   
32404816   Low spender                  8.0    Frequent customer   both   
32404817   Low spender                  8.0    Frequent customer   both   

           region customer_activity  
32404813  Midwest       High active  
32404814  Midwest       High active  
32404815  Midwest       High active  
32404816  Midwest       High active  
32404817  Midwest       High active

In [48]:
high_active_customers.shape

(30964564, 34)

In [49]:
high_active_customers.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30964564 entries, 0 to 32404817
Data columns (total 34 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   user_id                int32   
 1   gender                 object  
 2   state                  object  
 3   age                    int8    
 4   date_joined            object  
 5   number_of_dependants   int8    
 6   marital_status         object  
 7   income                 int64   
 8   order_id               int32   
 9   order_number           int8    
 10  orders_day_of_week     int8    
 11  order_hour_of_day      int8    
 12  days_since_last-order  float16 
 13  new_customer           bool    
 14  product_id             int32   
 15  add_to_cart_order      int32   
 16  reordered              int8    
 17  product_name           object  
 18  aisle_id               int8    
 19  department_id          int8    
 20  prices                 float16 
 21  price_range_loc        object

In [50]:
# Exporting subset 'high active customers'
high_active_customers.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'high_active_customers_5_and_more_orders.pkl'))